File: fsharp/all-your-base/AllYourBase.fs
module AllYourBase

    let private (|GreaterThan|_|) (a:'a when 'a:comparison) (b:'a when 'a:comparison) =
        match b > a with
        | true ->
            ()
            |> Some
        | false -> None

    let private (|ValidDigits|_|) (inputBase:int) (digits:int list) =
        match List.forall (fun (digit:int) -> digit > -1 && digit < inputBase) digits with
        | true ->
            ()
            |> Some
        | false -> None

    let private digitsToNumber (inputBase:int) (digits:int list):int =
        match digits with
        | [] -> 0
        | _ ->
            digits
            |> List.fold (fun (number:int) (digit:int) -> number * inputBase + digit) 0

    let private numberTodigits (outputBase:int) (number:int):int list =
        let rec tailCall (digits:int list) (remainder:int):int list =
            match remainder with
            | 0 -> digits
            | _ -> tailCall ((remainder % outputBase)::digits) (remainder / outputBase)
        let digits:int list = tailCall [] number
        match digits with
        | [] -> [0]
        | _ -> digits

    let rebase (digits:int list) (inputBase:int) (outputBase:int):int list option =
        match inputBase with
        | GreaterThan 1 ->
            match outputBase with
            | GreaterThan 1 ->
                match digits with
                | ValidDigits inputBase ->
                    digits
                    |> digitsToNumber inputBase
                    |> numberTodigits outputBase
                    |> Some
                | _ -> None
            | _ -> None
        | _ -> None

File: solutions/scala/all-your-base/2/src/main/scala/AllYourBase.scala
object AllYourBase
{

    private def ValidDigits(inputBase:Int, digits:List[Int]):Boolean = digits.forall((digit:Int) => digit > -1 && digit < inputBase)

    private def digitsToNumber(inputBase:Int, digits:List[Int]):Int =
    {
        digits match
        {
            case Nil => 0
            case _ => digits.foldLeft(0)((number:Int, digit:Int) => number * inputBase + digit) 
        }
    }

    private def numberTodigits(outputBase:Int, number:Int):List[Int] =
    {
        def tailCall (digits:List[Int]) (remainder:Int):List[Int] =
        {
            remainder match
            {
                case 0 => digits
                case _ => tailCall ((remainder % outputBase)::digits) (remainder / outputBase)
            }
        }
        def digits:List[Int] = tailCall (Nil) (number)
        digits match
        {
            case Nil => List(0)
            case _ => digits
        }
    }

    def rebase(inputBase:Int, digits:List[Int], outputBase:Int):Option[List[Int]] =
    {
        inputBase match
        {
            case inputBase if inputBase > 1 =>
                outputBase match
                case outputBase if outputBase > 1 =>
                    digits match
                    case digits if ValidDigits(inputBase, digits) =>
                        Some(numberTodigits(outputBase, digitsToNumber(inputBase, digits)))
                    case _ => None
                case _ => None
            case _ => None
        }
    }
}